mqtt-endpoint/src/service/session/mod.rs

```diff
@@ -108,33 +155,34 @@ where
         &self,
         publish: &Publish<'_>,
     ) -> Result<(String, Arc<registry::v1::Device>), PublishError> {
-        let topic = publish.topic().path().split('/').collect::<Vec<_>>();
-        log::debug!("Topic: {:?}", topic);
+        // TODO this should be removed and is just there as an example
 
-        match topic.as_slice() {
-            [channel] => Ok((channel.to_string(), self.device.clone())),
-            [channel, as_device] => self
-                .device_cache
-                .fetch(as_device, |as_device| {
-                    self.auth
-                        .authorize_as(
-                            &self.application.metadata.name,
-                            &self.device.metadata.name,
-                            as_device,
-                        )
-                        .map_ok(|result| match result.outcome {
-                            GatewayOutcome::Pass { r#as } => Some(r#as),
-                            _ => None,
-                        })
-                })
-                .await
-                .map(|r| (channel.to_string(), r)),
-            _ => Err(PublishError::TopicNameInvalid),
+        match &self.dialect.handle(publish.topic().path()) {
+            Ok(topic) => match topic.device {
```

Comment:
I think you can merge the two `match` layers into something like:

```rust
match &self.dialect.handle(...) {
  Ok(ParsedTopic{device: Some(device), channel}) => {},
  Ok(ParsedTopic{device: None, channel}) => {}
  Err(_) => {}
}
```

Reply:
Ok, I just tried that out. It works. But I doesn't make the code more readable. So let's keep it as it is.